lib/src/test/resources/http4s/form-params-017.txt

```diff
@@ -0,0 +1,113 @@
+    org.http4s.QueryParamDecoder.fromUnsafeCast[BigDecimal](p => BigDecimal(p.value))("BigDecimal")
+
+  implicit lazy val instantQueryParamDecoder: org.http4s.QueryParamDecoder[java.time.Instant] =
+    org.http4s.QueryParamDecoder.fromUnsafeCast[java.time.Instant](p => java.time.Instant.parse(p.value))("java.time.Instant")
+
+  implicit lazy val localDateQueryParamDecoder: org.http4s.QueryParamDecoder[java.time.LocalDate] =
+    org.http4s.QueryParamDecoder.fromUnsafeCast[java.time.LocalDate](p => java.time.LocalDate.parse(p.value))("java.time.LocalDate")
+
+  implicit lazy val uuidQueryParamDecoder: org.http4s.QueryParamDecoder[java.util.UUID] =
+    org.http4s.QueryParamDecoder.fromUnsafeCast[java.util.UUID](p => java.util.UUID.fromString(p.value))("java.util.UUID")
+
+
+  object ApiVersion {
+    val ApiVersionMajor = {
+      import org.http4s.syntax.string._
+      "X-Apidoc-Version-Major".ci
+    }
+
+    def apply(req: org.http4s.Message): Boolean = req.headers.get(ApiVersionMajor) match {
+      case Some(v) if v.value == "0" => true
+      case _ => false
+    }
+  }
+
+
+}
+
+trait ModelRoutes {
+  import Matchers._
+
+  implicit def circeJsonDecoder[A](implicit decoder: _root_.io.circe.Decoder[A]) = org.http4s.circe.jsonOf[A]
+  implicit def circeJsonEncoder[A](implicit encoder: _root_.io.circe.Encoder[A]) = org.http4s.circe.jsonEncoderOf[A]
+
+  case class PostRequest(id: String, model: io.apibuilder.http4s.test.models.Model, enum: io.apibuilder.http4s.test.models.Enum, opt: _root_.scala.Option[Int], list: Seq[String], map: Map[String, String])
+
+  implicit val PostRequestDecoder: _root_.io.circe.Decoder[PostRequest] = _root_.io.circe.Decoder.instance { a =>
+    for {
+      id <- a.downField("id").as[String]
+      model <- a.downField("model").as[io.apibuilder.http4s.test.models.Model]
+      enum <- a.downField("enum").as[io.apibuilder.http4s.test.models.Enum]
+      opt <- a.downField("opt").as[_root_.scala.Option[Int]]
+      list <- a.downField("list").as[Seq[String]]
+      map <- a.downField("map").as[Map[String, String]]
+    } yield {
+      PostRequest(
+        id = id,
+        model = model,
+        enum = enum,
+        opt = opt,
+        list = list,
+        map = map
+      )
+    }
+  }
+
+  sealed trait PostResponse
+
+  object PostResponse {
+    case class HTTP200(headers: Seq[org.http4s.Header] = Nil) extends PostResponse
+  }
+
+  def post(
+    _req: org.http4s.Request,
+    id: String,
+    model: io.apibuilder.http4s.test.models.Model,
+    enum: io.apibuilder.http4s.test.models.Enum,
+    opt: _root_.scala.Option[Int],
+    list: Seq[String],
+    map: Map[String, String]
+  ): fs2.Task[PostResponse]
+
+  def apiVersionMatch(req: org.http4s.Message): Boolean = ApiVersion(req)
+
+  def service() = org.http4s.HttpService {
+    case _req @ POST -> Root / "test" if apiVersionMatch(_req) =>
+    if (_req.contentType.exists(_.mediaType == _root_.org.http4s.MediaType.`application/json`)) {
```

Comment:
Is this actually possible? That a form POST is encoded as json?

Reply:
That's not a form POST but a generic POST. 
It's definitely possible to send "application/json" requests via curl, not sure if apibuilder clients can do it but should just be matter of specifying it in the headers I think.

Reply:
Yeah, but we have to take the apibuilder spec as the authority. And that should either have a [body declaration](https://app.apibuilder.io/doc/apiJson#body), in which case the request body contents are json encoded and we have an existing model type and de/serializer, *or* the operation has form params, in which case the request body is form encoded and we need to extract the parameters through org.http4s.UrlForm. But I don't think there should ever be a case (for an apibuilder generated operation) where a POST endpoint can be either form or json encoded.

Reply:
I think actually it would be quite nice to be able to support both intrinsically - and this is what we are doing at Flow btw. The APIs are documented as application/json with a body - but you can talk to the APIs with url-form-encoded if you like. Something has to know how to convert the url encoded to application/json - and there is an open question if this should happen in the apibuilder clients or elsewhere (at flow we do this in the reverse proxy)

Reply:
To me it would be nicer to handle the request either way, but I understand that it doesn't match the current spec. If you want I can remove the `application/json` related code from this PR (so at least we have a fix for `form-encoded` part) and discuss this separately.